### PR TITLE
feat: Better support of Gateway Subdomain spec

### PR DESCRIPTION
**Pull request by ppodolsky**
Added basic subdomain support by uplifting parameters extraction from `Host` header required for handlers.
DnsLink part of spec is not supported in this PR.

Related to #504 

**Spec:**

- [x] HEAD, GET methods
- [x] DNSLink support
- [x] Redirection from Path methods to Subdomain methods (behaviour is configured by option)
- [x] X-Forwarded-Host support
- [x] X-Forwarded-Proto support

Current PR won't cover [URI router](https://github.com/ipfs/specs/blob/main/http-gateways/SUBDOMAIN_GATEWAY.md#uri-router), I believe it is better to implement in a separate PR.

### CODE REVIEW

File: iroh-gateway/src/handlers.rs
Comment (dignifiedquire) [inline iroh-gateway/src/handlers.rs:58]: Is this a custom or based on somewhere?

Comment (dignifiedquire) [reply iroh-gateway/src/handlers.rs:58]: in either case, should have definitely some tests just making sure this captures the right things on different inputs

Comment (ppodolsky) [reply iroh-gateway/src/handlers.rs]: Added several tests

**Comment by ppodolsky** [comment]
Can't figure out if we should do anything here with the note from [spec](https://github.com/ipfs/specs/blob/main/http-gateways/SUBDOMAIN_GATEWAY.md):

> Note: Base36 must be used to ensure CIDv1 with ED25519 fits in a single DNS label (63 characters).

Decode and then encode IPNS name? Or what?

File: iroh-gateway/src/core.rs
Comment (lidel) [inline iroh-gateway/src/core.rs:521]: Would be nice to add separate test for legacy notation "peerid in Base58"  that was not a valid CID (e.g., `12D3KooWJHxkQKX8C5KAyqEPhn2ssT2in4TExyG9SXxi519tycL9`)
See explainer in https://github.com/multiformats/cid-utils-website/issues/48

This is important feature of subdomain gateway: when a request is made to `example.com/ipns/id`, subdomain set up on `example.com` will return HTTP 301 redirect to proper subdomain. 
Before redirect is made, `id` is converted to `cidv1-libp2p-key-base36`.

File: iroh-gateway/src/core.rs
Comment (lidel) [inline iroh-gateway/src/core.rs:557]: nit: change to `en.wikipedia-on-ipfs.org`? this is our usual test case for handling `-`  (i suggest avoiding using  .eth in tests unrelated to ENS interop).

File: iroh-gateway/src/core.rs
Comment (lidel) [inline iroh-gateway/src/core.rs:571]: nit: missing the same test for `x-forwarded-host`

File: iroh-gateway/src/regex.rs
Comment (lidel) [inline iroh-gateway/src/regex.rs:6]: Just to make sure there are no regressions in the future, add tests for edge cases where hostname is the same as scheme, e.g.:

`bafy.ipfs.ipfs.com` 
`bafy.ipns.ipfs.com`
`bafy.ipns.ipns.com`
`bafy.ipns.ipfs.com`

File: iroh-gateway/src/core.rs
Comment (ppodolsky) [reply iroh-gateway/src/core.rs:525]: Done

I had to add nasty `from_peer_id` function for dealing with discrepancy between `multihash:0.16` and `multihash:0.17`. This func may be safely deleted after updateing to multihash:0.17

cc @dignifiedquire Would you mind to release multiaddr 0.17 for bringing new multihash crate to rust-libp2p?

File: iroh-gateway/src/core.rs
Comment (ppodolsky) [reply iroh-gateway/src/core.rs:593]: Done

File: iroh-gateway/src/regex.rs
Comment (ppodolsky) [reply iroh-gateway/src/regex.rs]: Done

File: iroh-gateway/src/core.rs
Comment (ppodolsky) [reply iroh-gateway/src/core.rs]: Done.

Btw, while I was implementing code around `-` processing with regular expressions I had faced that it is hard or even impossible to properly process cases like `aaaa-b-aaaa.com` (a single symbol between 2 dashes) because captured groups were overlapping there (`a-b` and `b-a` with `b` belonging to both groups) and it is forbidden in `Regex` crate.

I mean, it may be useful to consider in E2E and other tests this case too cause it belongs to a separate equivalence class.

File: iroh-gateway/src/core.rs
Comment (dignifiedquire) [reply iroh-gateway/src/core.rs:525]: @vmx what't the status here?

Comment (vmx) [reply iroh-gateway/src/core.rs:525]: @dignifiedquire I'm not maintaining rust-multiaddr, that's a question for @mxinden :)

**Review by Arqu — COMMENTED** [review]
In regards to the gateway changes, this looks great. Regex is a bit scary looking.
Only thing I think we should do before the merge is resolve the multihash dependency.

File: iroh-resolver/src/resolver.rs
Comment (ppodolsky) [reply iroh-resolver/src/resolver.rs:50]: I need a moment to fix it, but we should decide who will release multiaddr 0.17 and bring it through rust-libp2p and possibly other deps

**Comment by ppodolsky** [comment]
> In regards to the gateway changes, this looks great. Regex is a bit scary looking. Only thing I think we should do before the merge is resolve the multihash dependency.

No need to scare, I replace regex with function :)

File: iroh-gateway/src/core.rs
Comment (lidel) [reply iroh-gateway/src/core.rs]: fwiw Kubo also started with regex, realized it comes with footguns, and replaced with a plain function :)

Comment (ppodolsky) [reply iroh-gateway/src/core.rs]: And it doesn't bring extra crate. You are right, dunno why I've decided regex would be better. It might be useful for complicated cases or expected future changes in format, but replacing dashes are definitely not this case.

File: iroh-gateway/src/core.rs
Comment (mxinden) [reply iroh-gateway/src/core.rs:525]: Kicked off release process here. Please approve. https://github.com/multiformats/rust-multiaddr/pull/66

**Comment by ppodolsky** [comment]
Almost done, but should not merge until I set proper version of libp2p in Cargo.toml

**Comment by ppodolsky** [comment]
We may create an issue for cleaning things out after libp2p release and merge this PR as is. If it is ok, I'll prepare proper commits tomorrow

**Comment by dignifiedquire** [comment]
> We may create an issue for cleaning things out after libp2p release and merge this PR as is. If it is ok, I'll prepare proper commits tomorrow

Yeah I think this makes the most sense, appreciated

**Comment by ppodolsky** [comment]
> > We may create an issue for cleaning things out after libp2p release and merge this PR as is. If it is ok, I'll prepare proper commits tomorrow
> 
> Yeah I think this makes the most sense, appreciated

Done

File: Cargo.toml
Comment (dignifiedquire) [inline Cargo.toml:50]: can you remove this from this PR here please, see discussion on #595

Comment (ppodolsky) [reply Cargo.toml]: Done

File: iroh-gateway/src/client.rs
Comment (Arqu) [inline iroh-gateway/src/client.rs:241]: Nitpick, can we make this `subdomain_mode`

File: iroh-gateway/src/handlers.rs
Comment (Arqu) [inline iroh-gateway/src/handlers.rs:57]: Can we rename this to `RespondImmediately` is a little bit more readable.

**Review by Arqu — CHANGES_REQUESTED** [review]
Just some very minor comments.
Need to get CI to pass, otherwise I'm happy to merge.

File: iroh-gateway/src/handlers.rs
Comment (ppodolsky) [reply iroh-gateway/src/handlers.rs]: Done

File: iroh-gateway/src/client.rs
Comment (ppodolsky) [reply iroh-gateway/src/client.rs]: Done

File: iroh-gateway/src/constants.rs
Comment (ppodolsky) [reply iroh-gateway/src/constants.rs:4]: Let's merge this one firstly and then I'll rebase and refine left one

File: iroh-gateway/src/config.rs
Comment (ppodolsky) [reply iroh-gateway/src/config.rs:53]: Done

**Comment by ppodolsky** [comment]
Argh, something happened during rebase, let me check what is wrong with the test

**Comment by ppodolsky** [comment]
Fixed tests

**Review by Arqu — APPROVED** [review]
LGTM!

**Pull request merged**
